// @tazama-lf/frms-coe-lib/lib/helpers/transactionTypeGuards.js
"use strict";
// SPDX-License-Identifier: Apache-2.0
Object.defineProperty(exports, "__esModule", { value: true });
exports.isBaseMessageTransaction = exports.isStructuredTransaction = exports.isPain013Transaction = exports.isPain001Transaction = exports.isPacs008Transaction = exports.isPacs002Transaction = void 0;
const isRecord = (value) => typeof value === 'object' && value !== null;
const isString = (value) => typeof value === 'string';
/**
 * Type guard to detect Pacs002 (pacs.002.001.12) transactions.
 * Checks for FIToFIPmtSts with GrpHdr and TxInfAndSts.
 */
const isPacs002Transaction = (transaction) => {
    if (!isRecord(transaction)) {
        return false;
    }
    if (transaction.TxTp !== 'pacs.002.001.12' || !isString(transaction.TenantId)) {
        return false;
    }
    const pacsRoot = transaction.FIToFIPmtSts;
    if (!isRecord(pacsRoot)) {
        return false;
    }
    return isRecord(pacsRoot.GrpHdr) && isRecord(pacsRoot.TxInfAndSts);
};
exports.isPacs002Transaction = isPacs002Transaction;
/**
 * Type guard to detect Pacs008 (pacs.008.001.10) transactions.
 * Checks for FIToFICstmrCdtTrf with GrpHdr and CdtTrfTxInf.
 */
const isPacs008Transaction = (transaction) => {
    if (!isRecord(transaction)) {
        return false;
    }
    if (transaction.TxTp !== 'pacs.008.001.10' || !isString(transaction.TenantId)) {
        return false;
    }
    const pacsRoot = transaction.FIToFICstmrCdtTrf;
    if (!isRecord(pacsRoot)) {
        return false;
    }
    return isRecord(pacsRoot.GrpHdr) && isRecord(pacsRoot.CdtTrfTxInf);
};
exports.isPacs008Transaction = isPacs008Transaction;
/**
 * Type guard to detect Pain001 (pain.001.001.11) transactions.
 * Checks for CstmrCdtTrfInitn with GrpHdr and PmtInf.
 */
const isPain001Transaction = (transaction) => {
    if (!isRecord(transaction)) {
        return false;
    }
    if (transaction.TxTp !== 'pain.001.001.11' || !isString(transaction.TenantId)) {
        return false;
    }
    const painRoot = transaction.CstmrCdtTrfInitn;
    if (!isRecord(painRoot)) {
        return false;
    }
    return isRecord(painRoot.GrpHdr) && isRecord(painRoot.PmtInf);
};
exports.isPain001Transaction = isPain001Transaction;
/**
 * Type guard to detect Pain013 (pain.013.001.09) transactions.
 * Checks for CdtrPmtActvtnReq with GrpHdr and PmtInf.
 */
const isPain013Transaction = (transaction) => {
    if (!isRecord(transaction)) {
        return false;
    }
    if (transaction.TxTp !== 'pain.013.001.09' || !isString(transaction.TenantId)) {
        return false;
    }
    const painRoot = transaction.CdtrPmtActvtnReq;
    if (!isRecord(painRoot)) {
        return false;
    }
    return isRecord(painRoot.GrpHdr) && isRecord(painRoot.PmtInf);
};
exports.isPain013Transaction = isPain013Transaction;
/**
 * Gate check: returns true if the transaction is any of the 4 supported structured types
 * (Pacs002, Pacs008, Pain001, Pain013). Use this before type-specific narrowing.
 */
const isStructuredTransaction = (transaction) => (0, exports.isPacs002Transaction)(transaction) ||
    (0, exports.isPacs008Transaction)(transaction) ||
    (0, exports.isPain001Transaction)(transaction) ||
    (0, exports.isPain013Transaction)(transaction);
exports.isStructuredTransaction = isStructuredTransaction;
/**
 * Type guard to detect BaseMessage transactions.
 */
const isBaseMessageTransaction = (transaction) => {
    if (!isRecord(transaction)) {
        return false;
    }
    if (!isString(transaction.TxTp) || !isString(transaction.TenantId) || !isString(transaction.MsgId)) {
        return false;
    }
    if (!isRecord(transaction.Payload)) {
        return false;
    }
    return !('FIToFIPmtSts' in transaction);
};
exports.isBaseMessageTransaction = isBaseMessageTransaction;
