// @tazama-lf/frms-coe-lib/lib/builders/configurationBuilder.js
"use strict";
// SPDX-License-Identifier: Apache-2.0
Object.defineProperty(exports, "__esModule", { value: true });
exports.configurationBuilder = configurationBuilder;
const tslib_1 = require("tslib");
const node_cache_1 = tslib_1.__importDefault(require("node-cache"));
const util = tslib_1.__importStar(require("node:util"));
const pg_1 = require("pg");
const utils_1 = require("../builders/utils");
const dbManager_1 = require("../services/dbManager");
const utils_2 = require("./utils");
async function configurationBuilder(manager, configurationConfig, cacheConfig, onConfigLoaded) {
    const conf = {
        host: configurationConfig.host,
        port: configurationConfig.port,
        database: configurationConfig.databaseName,
        user: configurationConfig.user,
        password: configurationConfig.password,
        ssl: (0, utils_2.getSSLConfig)(configurationConfig.certPath),
    };
    manager._configuration = new pg_1.Pool(conf);
    try {
        const dbReady = await (0, utils_1.isDatabaseReady)(manager._configuration);
        dbManager_1.readyChecks.ConfigurationDB = dbReady ? 'Ok' : 'err';
    }
    catch (error) {
        const err = error;
        dbManager_1.readyChecks.ConfigurationDB = `err, ${util.inspect(err)}`;
    }
    manager.nodeCache = cacheConfig?.localCacheEnabled ? new node_cache_1.default() : undefined;
    manager.getRuleConfig = async (ruleId, cfg, tenantId) => {
        const cacheKey = `${tenantId}_${ruleId}_${cfg}`;
        if (manager.nodeCache) {
            const cacheVal = manager.nodeCache.get(cacheKey);
            if (cacheVal) {
                return cacheVal;
            }
        }
        const query = {
            text: `SELECT 
                configuration
              FROM 
                rule 
              WHERE 
                ruleId = $1 
              AND 
                ruleCfg = $2
              AND
                tenantId = $3`,
            values: [ruleId, cfg, tenantId],
        };
        const queryRes = await manager._configuration.query(query);
        const toReturn = queryRes.rows.length > 0 ? queryRes.rows[0].configuration : undefined;
        if (toReturn) {
            onConfigLoaded?.(toReturn);
            if (manager.nodeCache) {
                manager.nodeCache.set(cacheKey, toReturn, cacheConfig?.localCacheTTL ?? 3000);
            }
        }
        return toReturn;
    };
    manager.getTypologyConfig = async (typologyId, typologyCfg, tenantId) => {
        const cacheKey = `${tenantId}_${typologyId}_${typologyCfg}`;
        if (manager.nodeCache) {
            const cacheVal = manager.nodeCache.get(cacheKey);
            if (cacheVal)
                return await Promise.resolve(cacheVal);
        }
        const query = {
            text: `SELECT
              configuration
            FROM
              typology
            WHERE
              typologyId = $1 
            AND 
              typologyCfg = $2
            AND
              tenantId = $3`,
            values: [typologyId, typologyCfg, tenantId],
        };
        const queryRes = await manager._configuration.query(query);
        const toReturn = queryRes.rows.length > 0 ? queryRes.rows[0].configuration : undefined;
        if (toReturn && manager.nodeCache) {
            manager.nodeCache.set(cacheKey, toReturn, cacheConfig?.localCacheTTL ?? 3000);
        }
        return toReturn;
    };
    manager.getNetworkMap = async () => {
        const query = {
            text: `SELECT
              configuration
            FROM
              network_map
            WHERE
              active = $1`,
            values: [true],
        };
        const queryRes = await manager._configuration.query(query);
        return queryRes.rows.length > 0 ? queryRes.rows.map((value) => value.configuration) : [];
    };
    manager.getPathPushJob = async (path, tenantId) => {
        const query = {
            text: 'SELECT * FROM tcs_push_jobs WHERE path = $1 AND tenant_id = $2 LIMIT 1;',
            values: [path, tenantId],
        };
        const queryRes = await manager._configuration.query(query);
        return queryRes.rows.length > 0 ? queryRes.rows[0] : undefined;
    };
    manager.getDefaultPushJob = async () => {
        const query = {
            text: 'SELECT * FROM tcs_push_jobs WHERE status IN($1, $2, $3) AND publishing_status = $4',
            values: ['STATUS_08_DEPLOYED', 'STATUS_06_EXPORTED', 'STATUS_04_APPROVED', 'active'],
        };
        const queryRes = await manager._configuration.query(query);
        return queryRes.rows;
    };
    manager.getIdPushJob = async (type, id) => {
        const text = type === 'push'
            ? `
          SELECT *
          FROM tcs_push_jobs
          WHERE id = $1
          LIMIT 1;
        `
            : `
        SELECT 
          j.*, 
           s.cron
            FROM tcs_pull_jobs j
             LEFT JOIN tcs_cron_jobs s ON j.schedule_id = s.id
              WHERE j.id = $1
               LIMIT 1;
        `;
        const query = {
            text,
            values: [id],
        };
        const queryRes = await manager._configuration.query(query);
        return queryRes.rows.length > 0 ? queryRes.rows[0] : undefined;
    };
    manager.insertJobHistory = async (tenantId, jobId, counts, processedCounts, exception, jobType) => {
        const query = {
            text: 'INSERT INTO job_history (tenant_id, job_id, counts, processed_counts, exception, job_type) VALUES ($1, $2, $3, $4, $5, $6);',
            values: [tenantId, jobId, counts, processedCounts, exception, jobType],
        };
        await manager._configuration.query(query);
    };
}
