// @tazama-lf/frms-coe-lib/lib/builders/enrichmentBuilder.js
"use strict";
// SPDX-License-Identifier: Apache-2.0
Object.defineProperty(exports, "__esModule", { value: true });
exports.enrichmentBuilder = enrichmentBuilder;
const tslib_1 = require("tslib");
const util = tslib_1.__importStar(require("node:util"));
const pg_1 = require("pg");
const utils_1 = require("../builders/utils");
const dbManager_1 = require("../services/dbManager");
const utils_2 = require("./utils");
async function enrichmentBuilder(manager, enrichmentConfig) {
    const conf = {
        host: enrichmentConfig.host,
        port: enrichmentConfig.port,
        database: enrichmentConfig.databaseName,
        user: enrichmentConfig.user,
        password: enrichmentConfig.password,
        ssl: (0, utils_2.getSSLConfig)(enrichmentConfig.certPath),
    };
    manager._enrichment = new pg_1.Pool(conf);
    try {
        const dbReady = await (0, utils_1.isDatabaseReady)(manager._enrichment);
        dbManager_1.readyChecks.EnrichmentDB = dbReady ? 'Ok' : 'err';
    }
    catch (err) {
        dbManager_1.readyChecks.EnrichmentDB = `err, ${util.inspect(err)}`;
    }
    manager.ingestData = async (text, values = []) => {
        const query = {
            text,
            values,
        };
        await manager._enrichment.query(query);
    };
    manager.createTable = async (tableName) => {
        // Validate table name to prevent SQL injection
        const validatedTableName = (0, utils_1.validateTableName)(tableName);
        const createQuery = `
        CREATE TABLE IF NOT EXISTS ${validatedTableName} (
          id UUID PRIMARY KEY DEFAULT gen_random_uuid(),
          data JSONB NOT NULL,
          job_id TEXT NOT NULL,
          checksum TEXT NOT NULL,
          created_at TIMESTAMP NOT NULL DEFAULT NOW()
        );
      `;
        const query = {
            text: createQuery,
            values: [],
        };
        await manager._enrichment.query(query);
    };
    manager.deleteRows = async (tableName) => {
        // Validate table name to prevent SQL injection
        const validatedTableName = (0, utils_1.validateTableName)(tableName);
        const query = {
            text: `
       DELETE FROM ${validatedTableName};
      `,
            values: [],
        };
        await manager._enrichment.query(query);
    };
}
