// @tazama-lf/frms-coe-lib/lib/builders/rawHistoryBuilder.js
"use strict";
// SPDX-License-Identifier: Apache-2.0
Object.defineProperty(exports, "__esModule", { value: true });
exports.rawHistoryBuilder = rawHistoryBuilder;
const tslib_1 = require("tslib");
const util = tslib_1.__importStar(require("node:util"));
const pg_1 = require("pg");
const pg_format_1 = tslib_1.__importDefault(require("pg-format"));
const utils_1 = require("../builders/utils");
const dbManager_1 = require("../services/dbManager");
const utils_2 = require("./utils");
const MAX_PAYLOAD_SIZE = 1024 * 1024; // 1MB
async function rawHistoryBuilder(manager, rawHistoryConfig) {
    const conf = {
        host: rawHistoryConfig.host,
        port: rawHistoryConfig.port,
        database: rawHistoryConfig.databaseName,
        user: rawHistoryConfig.user,
        password: rawHistoryConfig.password,
        ssl: (0, utils_2.getSSLConfig)(rawHistoryConfig.certPath),
    };
    manager._rawHistory = new pg_1.Pool(conf);
    try {
        const dbReady = await (0, utils_1.isDatabaseReady)(manager._rawHistory);
        dbManager_1.readyChecks.RawHistoryDB = dbReady ? 'Ok' : 'err';
    }
    catch (error) {
        const err = error;
        dbManager_1.readyChecks.RawHistoryDB = `err, ${util.inspect(err)}`;
    }
    manager.getTransactionPacs008 = async (endToEndId, tenantId) => {
        const query = {
            text: 'SELECT document FROM pacs008 WHERE endToEndId = $1 AND tenantId = $2',
            values: [endToEndId, tenantId],
        };
        const queryRes = await manager._rawHistory.query(query);
        const toReturn = queryRes.rows.length > 0 ? queryRes.rows[0].document : undefined;
        return toReturn;
    };
    manager.saveTransactionHistoryPain001 = async (tran) => {
        const query = {
            text: 'INSERT INTO pain001 (document) VALUES ($1) ON CONFLICT (EndToEndId, tenantId) DO NOTHING',
            values: [tran],
        };
        await manager._rawHistory.query(query);
    };
    manager.saveTransactionHistoryPain013 = async (tran) => {
        const query = {
            text: 'INSERT INTO pain013 (document) VALUES ($1) ON CONFLICT (EndToEndId, tenantId) DO NOTHING',
            values: [tran],
        };
        await manager._rawHistory.query(query);
    };
    manager.saveTransactionHistoryPacs008 = async (tran) => {
        const query = {
            text: 'INSERT INTO pacs008 (document) VALUES ($1) ON CONFLICT (EndToEndId, tenantId) DO NOTHING',
            values: [tran],
        };
        await manager._rawHistory.query(query);
    };
    manager.saveTransactionHistoryPacs002 = async (tran) => {
        const query = {
            text: 'INSERT INTO pacs002 (document) VALUES ($1) ON CONFLICT (EndToEndId, tenantId) DO NOTHING',
            values: [tran],
        };
        await manager._rawHistory.query(query);
    };
    // ---------------------
    manager.saveToQuarantine = async (record) => {
        let payload = record.raw_payload;
        if (payload.length > MAX_PAYLOAD_SIZE) {
            payload = payload.substring(0, MAX_PAYLOAD_SIZE) + '... [truncated]';
        }
        const quarantineQuery = {
            text: 'INSERT INTO dems_quarantine (id, correlation_id, tenant_id, endpoint_path, config_id, version, error, raw_payload, status) VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9)',
            values: [
                record.id,
                record.correlation_id,
                record.tenant_id,
                record.endpoint_path,
                record.config_id,
                record.version,
                record.error,
                payload,
                record.status,
            ],
        };
        await manager._rawHistory.query(quarantineQuery);
    };
    manager.saveDynamicTransactionHistory = async (tableName, tran, trackedFields) => {
        if (!/^[a-zA-Z_][a-zA-Z0-9_]{0,62}$/.test(tableName)) {
            throw new Error(`Invalid table name format: ${tableName}. Table names must start with a letter or underscore and contain only letters, digits, and underscores (max 63 characters).`);
        }
        if (!trackedFields?.TenantId) {
            throw new Error('TenantId is required for transaction history - essential for data isolation');
        }
        if (!trackedFields?.CreDtTm) {
            throw new Error('CreDtTm (creation date/time) is required for transaction history - essential for audit trail');
        }
        const randomMessageId = `msg-${Math.random().toString(36).substring(2, 15)}`;
        const query = {
            text: (0, pg_format_1.default)('INSERT INTO %I (document, credttm, messageid, endtoendid, debtoraccountid, creditoraccountid, tenantid) VALUES ($1, $2, $3, $4, $5, $6, $7)', tableName),
            values: [
                tran,
                trackedFields.CreDtTm,
                trackedFields.MsgId ?? randomMessageId,
                trackedFields.EndToEndId?.trim() ?? trackedFields.MsgId?.trim(),
                trackedFields.dbtrAcctId ?? null,
                trackedFields.cdtrAcctId ?? null,
                trackedFields.TenantId.trim(),
            ],
        };
        await manager._rawHistory.query(query);
    };
    manager.getTransactionAny = async (endToEndId, tenantId, tableName) => {
        if (!/^[a-zA-Z_][a-zA-Z0-9_]{0,62}$/.test(tableName)) {
            throw new Error(`Invalid table name format: ${tableName}. Table names must start with a letter or underscore and contain only letters, digits, and underscores (max 63 characters).`);
        }
        const query = {
            text: (0, pg_format_1.default)('SELECT document FROM %I WHERE endToEndId = $1 AND tenantId = $2', tableName),
            values: [endToEndId.trim(), tenantId.trim()],
        };
        const queryRes = await manager._rawHistory.query(query);
        const toReturn = queryRes.rows.length > 0 ? queryRes.rows[0].document : undefined;
        return toReturn;
    };
}
