// @tazama-lf/frms-coe-lib/lib/schemas/ruleConfig.js
"use strict";
// SPDX-License-Identifier: Apache-2.0
Object.defineProperty(exports, "__esModule", { value: true });
exports.baseRuleConfigSchema = exports.baseConfigSchema = exports.CaseSchema = exports.ExpressionSchema = exports.BandSchema = exports.OutcomeResultSchema = void 0;
const zod_1 = require("zod");
exports.OutcomeResultSchema = zod_1.z.object({
    subRuleRef: zod_1.z.string(),
    reason: zod_1.z.string(),
});
exports.BandSchema = exports.OutcomeResultSchema.extend({
    lowerLimit: zod_1.z.number().optional(),
    upperLimit: zod_1.z.number().optional(),
});
exports.ExpressionSchema = exports.OutcomeResultSchema.extend({
    value: zod_1.z.string(),
});
exports.CaseSchema = zod_1.z.object({
    expressions: zod_1.z.array(exports.ExpressionSchema),
    alternative: exports.OutcomeResultSchema,
});
exports.baseConfigSchema = zod_1.z.object({
    parameters: zod_1.z.record(zod_1.z.union([zod_1.z.string(), zod_1.z.number()]), zod_1.z.unknown()).optional(),
    exitConditions: zod_1.z.array(exports.OutcomeResultSchema).optional(),
    bands: zod_1.z.array(exports.BandSchema).optional(),
    cases: exports.CaseSchema.optional(),
});
exports.baseRuleConfigSchema = zod_1.z.object({
    id: zod_1.z.string(),
    cfg: zod_1.z.string(),
    tenantId: zod_1.z.string(),
    desc: zod_1.z.string().optional(),
    creDtTm: zod_1.z.string().optional(),
    updDtTm: zod_1.z.string().optional(),
    config: exports.baseConfigSchema,
});
